File: python/src/kyccentral/resources/registries.py
"""Sector registries and jurisdiction reference data."""

from __future__ import annotations

from typing import Any

from .._transport import AsyncResource, SyncResource, _seg

__all__ = [
    "AsyncCharity",
    "AsyncHmrcVat",
    "AsyncJurisdictions",
    "AsyncOffshoreJurisdictions",
    "Charity",
    "HmrcVat",
    "Jurisdictions",
    "OffshoreJurisdictions",
]

JSON = dict[str, Any]


class Charity(SyncResource):
    """Charity Commission for England and Wales lookups."""

    def status(self) -> JSON:
        """Whether Charity Commission access is configured."""
        return self._get("/charity/status")

    def search(self, q: str) -> JSON:
        """Search registered charities by name."""
        return self._get("/charity/search", params={"q": q})

    def get(self, registration_number: str, *, suffix: str | None = None) -> JSON:
        """A charity's register entry.

        Args:
            registration_number: Charity registration number.
            suffix: Subsidiary suffix, for linked charities registered under one
                number. Defaults to the parent entry.
        """
        return self._get(
            f"/charity/charity/{_seg(registration_number, 'registration_number')}",
            params={"suffix": suffix},
        )

    def trustees(self, registration_number: str) -> JSON:
        """The charity's registered trustees."""
        return self._get(
            f"/charity/charity/{_seg(registration_number, 'registration_number')}/trustees"
        )


class AsyncCharity(AsyncResource):
    """Awaitable mirror of :class:`Charity`."""

    async def status(self) -> JSON:
        """Whether Charity Commission access is configured."""
        return await self._get("/charity/status")

    async def search(self, q: str) -> JSON:
        """Search registered charities by name."""
        return await self._get("/charity/search", params={"q": q})

    async def get(self, registration_number: str, *, suffix: str | None = None) -> JSON:
        """A charity's register entry."""
        return await self._get(
            f"/charity/charity/{_seg(registration_number, 'registration_number')}",
            params={"suffix": suffix},
        )

    async def trustees(self, registration_number: str) -> JSON:
        """The charity's registered trustees."""
        return await self._get(
            f"/charity/charity/{_seg(registration_number, 'registration_number')}/trustees"
        )


class HmrcVat(SyncResource):
    """HMRC VAT registration lookups."""

    def status(self) -> JSON:
        """Whether HMRC VAT lookups are configured."""
        return self._get("/hmrc-vat/status")

    def check(self, vat_number: str) -> JSON:
        """Check a UK VAT registration number.

        Args:
            vat_number: VAT number, with or without the ``GB`` prefix.
        """
        return self._get("/hmrc-vat/check", params={"vat_number": vat_number})


class AsyncHmrcVat(AsyncResource):
    """Awaitable mirror of :class:`HmrcVat`."""

    async def status(self) -> JSON:
        """Whether HMRC VAT lookups are configured."""
        return await self._get("/hmrc-vat/status")

    async def check(self, vat_number: str) -> JSON:
        """Check a UK VAT registration number."""
        return await self._get("/hmrc-vat/check", params={"vat_number": vat_number})


class Jurisdictions(SyncResource):
    """FATF high-risk and increased-monitoring jurisdiction reference data.

    Updated after each FATF plenary (roughly February, June and October).
    """

    def list(self) -> Any:
        """The current FATF black list and grey list."""
        return self._get("/jurisdictions")

    def check(self, country: str) -> JSON:
        """Whether one country is FATF-listed, and on which list."""
        return self._get("/jurisdictions/check", params={"country": country})


class AsyncJurisdictions(AsyncResource):
    """Awaitable mirror of :class:`Jurisdictions`."""

    async def list(self) -> Any:
        """The current FATF black list and grey list."""
        return await self._get("/jurisdictions")

    async def check(self, country: str) -> JSON:
        """Whether one country is FATF-listed, and on which list."""
        return await self._get("/jurisdictions/check", params={"country": country})


class OffshoreJurisdictions(SyncResource):
    """Offshore and shell-company jurisdiction reference data."""

    def list(self) -> Any:
        """Known offshore jurisdictions and the keywords used to detect them."""
        return self._get("/offshore-jurisdictions")

    def check(self, name: str) -> JSON:
        """Whether a name or address indicates an offshore jurisdiction."""
        return self._get("/offshore-jurisdictions/check", params={"name": name})


class AsyncOffshoreJurisdictions(AsyncResource):
    """Awaitable mirror of :class:`OffshoreJurisdictions`."""

    async def list(self) -> Any:
        """Known offshore jurisdictions and their detection keywords."""
        return await self._get("/offshore-jurisdictions")

    async def check(self, name: str) -> JSON:
        """Whether a name or address indicates an offshore jurisdiction."""
        return await self._get("/offshore-jurisdictions/check", params={"name": name})

File: javascript/src/resources/registries.ts
/** Sector registries and jurisdiction reference data. */

import { seg, type Transport } from '../transport.js';
import type { JsonObject } from '../types.js';
import type { CallOptions } from './companies.js';

export interface CharityOptions extends CallOptions {
  /** Subsidiary suffix, for linked charities registered under one number. */
  suffix?: string;
}

/** Charity Commission for England and Wales lookups. */
export class Charity {
  constructor(private readonly transport: Transport) {}

  /** Whether Charity Commission access is configured. */
  async status(options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/charity/status', options);
  }

  /** Search registered charities by name. */
  async search(q: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/charity/search', { params: { q }, signal: options.signal });
  }

  /** A charity's register entry. */
  async get(registrationNumber: string, options: CharityOptions = {}): Promise<JsonObject> {
    return this.transport.get(`/charity/charity/${seg(registrationNumber, 'registrationNumber')}`, {
      params: { suffix: options.suffix },
      signal: options.signal,
    });
  }

  /** The charity's registered trustees. */
  async trustees(registrationNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get(
      `/charity/charity/${seg(registrationNumber, 'registrationNumber')}/trustees`,
      options,
    );
  }
}

/** HMRC VAT registration lookups. */
export class HmrcVat {
  constructor(private readonly transport: Transport) {}

  /** Whether HMRC VAT lookups are configured. */
  async status(options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/hmrc-vat/status', options);
  }

  /** Check a UK VAT registration number, with or without the `GB` prefix. */
  async check(vatNumber: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/hmrc-vat/check', {
      params: { vat_number: vatNumber },
      signal: options.signal,
    });
  }
}

/**
 * FATF high-risk and increased-monitoring jurisdiction reference data.
 *
 * Updated after each FATF plenary — roughly February, June and October.
 */
export class Jurisdictions {
  constructor(private readonly transport: Transport) {}

  /** The current FATF black list and grey list. */
  async list(options: CallOptions = {}): Promise<unknown> {
    return this.transport.get('/jurisdictions', options);
  }

  /** Whether one country is FATF-listed, and on which list. */
  async check(country: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/jurisdictions/check', {
      params: { country },
      signal: options.signal,
    });
  }
}

/** Offshore and shell-company jurisdiction reference data. */
export class OffshoreJurisdictions {
  constructor(private readonly transport: Transport) {}

  /** Known offshore jurisdictions and the keywords used to detect them. */
  async list(options: CallOptions = {}): Promise<unknown> {
    return this.transport.get('/offshore-jurisdictions', options);
  }

  /** Whether a name or address indicates an offshore jurisdiction. */
  async check(name: string, options: CallOptions = {}): Promise<JsonObject> {
    return this.transport.get('/offshore-jurisdictions/check', {
      params: { name },
      signal: options.signal,
    });
  }
}
